src/dune_engine/dep.ml

```diff
@@ -315,44 +314,44 @@ module Set = struct
     file_selector
       (File_selector.create ~dir File_selector.Predicate_with_id.false_)
 
-  module Source_tree_map_reduce =
-    Source_tree.Dir.Make_map_reduce (Memo) (Monoid.Union (M))
-
-  let source_tree dir =
-    let prefix_with, dir = Path.extract_build_context_dir_exn dir in
-    Source_tree.find_dir dir >>= function
-    | None -> Memo.return empty
-    | Some dir ->
-      Source_tree_map_reduce.map_reduce dir ~traverse:Sub_dirs.Status.Set.all
-        ~f:(fun dir ->
-          let files = Source_tree.Dir.files dir in
-          let path =
-            Path.append_source prefix_with (Source_tree.Dir.path dir)
-          in
+  module Source_tree
+      (Monad : Monad.S) (Union : sig
+        val union_all :
+             Path.t
+          -> f:(path:Path.t -> files:Filename.Set.t -> t Monad.t)
+          -> t Monad.t
+      end) =
+  struct
+    let files dir =
+      Union.union_all dir ~f:(fun ~path ~files ->
+          Monad.return
```

Comment:
Wait, why do we ask for a monadic call back then if we are calling it with essentially a pure function?

Reply:
Seems like we don't need a monad in this API at all?

Reply:
Good question. It's because of the subsequent definition of:

```
    let files_with_set dir =
      let open Monad.O in
      let+ t = files dir in
      let paths =
        fold t ~init:Path.Set.empty ~f:(fun dep acc ->
            match dep with
            | File f -> Path.Set.add acc f
            | File_selector fs ->
              assert (
                File_selector.Predicate_with_id.equal
                  (File_selector.predicate fs)
                  File_selector.Predicate_with_id.false_);
              acc
            | _ -> assert false)
      in
      (t, paths)
```

Which requires analyzing the result returned by `files`. Since we must always traverse the source tree monadically (with memo in practice), every traversal must return the monadic value. I could remove the monad from the callback though.

Reply:
Perhaps I could also move `files_with_set` to the rules if it doesn't seem important enough?

Reply:
I think we should just extract `to_file_exn` bit  from `files_with_set` (which is the only interesting bit of code) to a separate function in this module, and drop `files_with_set`.

Reply:
Okay, I simplified it.

Looking at this again, `to_file_set_exn` seems like quite an ugly function to have in the engine API. It depends on the predicate_lang stuff, and for that reason I think it should be moved to its callers.

I suppose this can be done in a subsequent PR.